hashintel/hash
# Support related errors in `error-stack`

## 🌟 What is the purpose of this PR?

This extends the `error-stack` implementation to support multiple frames and sources, and also removes the `ReportImpl` in favor of an attachment-based approach.

## 🔗 Related links

* [Discussion](https://github.com/hashintel/hash/discussions/737)


## 🚫 Blocked by

* #777 

## 🏃 Follow Up

The initial PR has been split up into three parts:

1) implementation (this one)
2) hooks (new output in ASCII, fancy and json) (not yet opened)
3) #777 

## 🔍 What does this change?

* merge `ReportImpl` and `Report`
* change from single frames to frame


## 📜 Does this require a change to the docs?

Yes

* adds breaking change
* possibility of multiple sources

## ⚠️ Known issues

## 🐾 Next steps

* [x] merge `ReportImpl` and `Report`
* [x] add multiple sources to `Report` and `Frame`
* [x] common trait implementation for `Report`, this includes: `Extend`, `FromIterator`, ...
* [x] update RequestRef and co
* [x] change rendering logic
* [x] write up documentation


## 🛡 What tests cover this?

* unit tests
* integration tests for the formatting

## ❓ How to test this?

1.  Checkout the branch / view the deployment
2.  Use as a library in a binary crate
3.  Create multiple errors
4.  Merge those errors and see the output or query for information

## 📹 Demo

The output shown here has been deferred to another PR, that will be opened after this one.

<img width="907" alt="image" src="https://user-images.githubusercontent.com/7252775/177989206-4c465a36-24c8-4495-8e62-4911e73739f9.png">

<img width="241" alt="image" src="https://user-images.githubusercontent.com/7252775/177989167-113ed4a8-f3a9-4b0a-94df-01b0e7fc8e7f.png">

<img width="747" alt="image" src="https://user-images.githubusercontent.com/7252775/177989344-0c0bfa93-27ff-4c8d-90ee-6c191e41349d.png">


(without the fancy feature flag)
